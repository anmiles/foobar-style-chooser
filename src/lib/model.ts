import ko from 'knockout';
import type { Sections } from '../types';
import { Section } from './section';

export { Model };

class Model {
	sections: Section[];
	toString: ko.Computed<string>;

	constructor() {
		const sectionsJSON = this.getSections();
		this.sections      = sectionsJSON.map(({ title, styles }) => new Section(title, styles));

		this.toString = ko.computed(() => `:${this.sections
			.map(
				(section) => section.styles
					.filter((style) => style.enabled())
					.map((style) => style.title),
			)
			.flat()
			.sort()
			.join(':')}:`);

	}

	private getSections(): Sections {
		return require('../../sections.json');
	}
}
