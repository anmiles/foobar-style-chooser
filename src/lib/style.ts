import ko from 'knockout';

class Style {
	enabled : ko.Observable<boolean>;

	constructor(
		public title: string,
	) {
		this.enabled = ko.observable(false);
	}

	toggle(): void {
		this.enabled(!this.enabled());
	}
}

export { Style };
