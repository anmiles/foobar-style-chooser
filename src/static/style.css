* {
	box-sizing: border-box;
}

html {
	overflow-y: scroll;
}

html, body {
	width: 100%;
	min-height: 100%;
	margin: 0;
	padding: 0;
}

body {
	font-size: 20px;
}

body, input, button {
	font-family: Arial, sans-serif;
}

.main {
    width: 100%;
    min-height: 100%;
    flex-direction: column;
    display: flex;
    gap: 16px;
    padding: 16px;
}

.section {
    flex-direction: row;
    flex-wrap: wrap;
    border: 2px solid currentColor;
    border-radius: 1em;
    position: relative;
    display: grid;
    grid-column-gap: 16px;
    grid-row-gap: 16px;
    padding: 16px;
    grid-template-columns: repeat(12, minmax(0, 1fr));
}

.section {
	color: hsl(20, 50%, 50%);
}

.section.genres {
	color: hsl(120, 50%, 50%);
}

.section.instruments {
	color: hsl(200, 50%, 50%);
}

.section.hidden {
	display: none;
}

.section.output {
	color: hsl(0, 0%, 50%);
}

.section:after {
	content: '';
    display: block;
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    border-radius: inherit;
    background: currentColor;
    filter: opacity(0.3);
}

.button {
    text-transform: capitalize;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    font-size: 1em;
    color: inherit;
    background: rgba(255, 255, 255, 0.25);
    border: 1px solid currentColor;
    border-radius: 0.5em;
    position: relative;
    z-index: 1;
    outline: none;
    padding: 16px;
}

.button:hover {
    background: rgba(255, 255, 255, 0.5);
}

.button.enabled,
.button.enabled:hover {
	background: white;
}

.section.output {
    display: block;
}

.section.output .button {
    background: white;
	width: 100%;
	cursor: default;
}
