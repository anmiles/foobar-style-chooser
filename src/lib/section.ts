import { Style } from './style';

export { Section };

class Section {
	title: string;
	hidden: boolean;
	styles: Style[];

	constructor({ title, hidden, styles } : { title: string, hidden?: boolean, styles: string[] }) {
		this.title  = title;
		this.hidden = hidden || false;
		this.styles = styles.map((style) => new Style(style));
	}
}
