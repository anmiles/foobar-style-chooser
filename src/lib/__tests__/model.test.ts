import fs from 'fs';
import { Model } from '../model';
import { Section } from '../section';
import { Style } from '../style';

const enabledStyles: Record<string, boolean> = {
	style1 : true,
	style2 : false,
	style3 : true,
	style4 : false,
};

function createModel(): { model: Model, allStyles: Style[] } {
	const model = new Model();

	model.sections = [
		new Section({ title : 'section1', styles : [ 'style1', 'style2' ] }),
		new Section({ title : 'section2', styles : [ 'style3', 'style4' ] }),
	];

	const allStyles = model.sections.map((section) => section.styles).flat();

	allStyles.forEach((style) => {
		const enabled = enabledStyles[style.title];

		if (typeof enabled === 'undefined') {
			throw `Cannot index 'enabledStyles' with ${style.title}`;
		}

		style.enabled(enabled);
	});

	return {
		model,
		allStyles,
	};
}

if (!fs.existsSync('sections.json')) {
	fs.cpSync('sections.sample.json', 'sections.json');
}

jest.doMock('../../../sections.json', () => require('../../../sections.sample.json'));

describe('src/lib/model', () => {
	it('should construct model with 2 sections', () => {
		const model = new Model();

		expect(model).toMatchSnapshot();
	});

	describe('value', () => {
		describe('read', () => {
			it('should return joined result of section.value', () => {
				const { model } = createModel();

				expect(model.value()).toEqual(':style1:style3:');
			});

			it('should sort styles alphabetically', () => {
				const { model } = createModel();

				expect(model.value()).toEqual(':style1:style3:');
			});
		});

		describe('write', () => {
			it('should make styles enabled based on splitted value', () => {
				const { model, allStyles } = createModel();

				model.value(':style2:style4:');

				const enabledStyles = allStyles
					.filter((style) => style.enabled())
					.map((style) => style.title);

				expect(model.value()).toEqual(':style2:style4:');
				expect(enabledStyles).toEqual([ 'style2', 'style4' ]);
			});

			it('should sort styles alphabetically', () => {
				const { model, allStyles } = createModel();

				model.value(':style4:style2:');

				const enabledStyles = allStyles
					.filter((style) => style.enabled())
					.map((style) => style.title);

				expect(model.value()).toEqual(':style2:style4:');
				expect(enabledStyles).toEqual([ 'style2', 'style4' ]);
			});
		});

		describe('input', () => {
			it('should call parse function with string value', () => {
				const { model } = createModel();
				const parseSpy  = jest.spyOn(model as any, 'parse');

				const value = 'value';
				const event = { target : { value } };

				model['input'](model, event);

				expect(parseSpy).toHaveBeenCalledWith(value);
			});
		});

	});
});
