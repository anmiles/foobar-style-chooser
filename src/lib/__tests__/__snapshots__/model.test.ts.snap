// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/lib/model should construct model with 2 sections 1`] = `
Model {
  "sections": [
    Section {
      "hidden": false,
      "styles": [
        Style {
          "enabled": [Function],
          "title": "8bit",
        },
        Style {
          "enabled": [Function],
          "title": "Ambient",
        },
        Style {
          "enabled": [Function],
          "title": "Avantgarde",
        },
        Style {
          "enabled": [Function],
          "title": "Classical",
        },
        Style {
          "enabled": [Function],
          "title": "Downtempo",
        },
        Style {
          "enabled": [Function],
          "title": "Dub",
        },
        Style {
          "enabled": [Function],
          "title": "Dubstep",
        },
        Style {
          "enabled": [Function],
          "title": "Ethnic",
        },
        Style {
          "enabled": [Function],
          "title": "Hardcore",
        },
        Style {
          "enabled": [Function],
          "title": "House",
        },
        Style {
          "enabled": [Function],
          "title": "Jazz",
        },
        Style {
          "enabled": [Function],
          "title": "Jungle",
        },
        Style {
          "enabled": [Function],
          "title": "Latin",
        },
        Style {
          "enabled": [Function],
          "title": "Metal",
        },
        Style {
          "enabled": [Function],
          "title": "NewAge",
        },
        Style {
          "enabled": [Function],
          "title": "NuMetal",
        },
        Style {
          "enabled": [Function],
          "title": "Pop",
        },
        Style {
          "enabled": [Function],
          "title": "Psy",
        },
        Style {
          "enabled": [Function],
          "title": "Punk",
        },
        Style {
          "enabled": [Function],
          "title": "Rap",
        },
        Style {
          "enabled": [Function],
          "title": "Reggae",
        },
        Style {
          "enabled": [Function],
          "title": "Retro",
        },
        Style {
          "enabled": [Function],
          "title": "Rock",
        },
        Style {
          "enabled": [Function],
          "title": "RockBallad",
        },
        Style {
          "enabled": [Function],
          "title": "RocknRoll",
        },
        Style {
          "enabled": [Function],
          "title": "Sacred",
        },
        Style {
          "enabled": [Function],
          "title": "Soundtrack",
        },
        Style {
          "enabled": [Function],
          "title": "Southern",
        },
        Style {
          "enabled": [Function],
          "title": "Trance",
        },
      ],
      "title": "genres",
    },
    Section {
      "hidden": false,
      "styles": [
        Style {
          "enabled": [Function],
          "title": "Choir",
        },
        Style {
          "enabled": [Function],
          "title": "Drive",
        },
        Style {
          "enabled": [Function],
          "title": "Electronic",
        },
        Style {
          "enabled": [Function],
          "title": "Guitar",
        },
        Style {
          "enabled": [Function],
          "title": "Instrumental",
        },
        Style {
          "enabled": [Function],
          "title": "Nature",
        },
        Style {
          "enabled": [Function],
          "title": "Orchestra",
        },
        Style {
          "enabled": [Function],
          "title": "Piano",
        },
        Style {
          "enabled": [Function],
          "title": "Vargan",
        },
        Style {
          "enabled": [Function],
          "title": "Violin",
        },
      ],
      "title": "instruments",
    },
    Section {
      "hidden": false,
      "styles": [
        Style {
          "enabled": [Function],
          "title": "Active",
        },
        Style {
          "enabled": [Function],
          "title": "Atmosphere",
        },
        Style {
          "enabled": [Function],
          "title": "Chant",
        },
        Style {
          "enabled": [Function],
          "title": "Children",
        },
        Style {
          "enabled": [Function],
          "title": "Coffee",
        },
        Style {
          "enabled": [Function],
          "title": "Dance",
        },
        Style {
          "enabled": [Function],
          "title": "Depress",
        },
        Style {
          "enabled": [Function],
          "title": "Deutsch",
        },
        Style {
          "enabled": [Function],
          "title": "Festival",
        },
        Style {
          "enabled": [Function],
          "title": "Happy",
        },
        Style {
          "enabled": [Function],
          "title": "History",
        },
        Style {
          "enabled": [Function],
          "title": "Holiday",
        },
        Style {
          "enabled": [Function],
          "title": "Keys",
        },
        Style {
          "enabled": [Function],
          "title": "Long",
        },
        Style {
          "enabled": [Function],
          "title": "Love",
        },
        Style {
          "enabled": [Function],
          "title": "Massage",
        },
        Style {
          "enabled": [Function],
          "title": "PlayCalimba",
        },
        Style {
          "enabled": [Function],
          "title": "PlayGuitar",
        },
        Style {
          "enabled": [Function],
          "title": "PlayPiano",
        },
        Style {
          "enabled": [Function],
          "title": "Relax",
        },
        Style {
          "enabled": [Function],
          "title": "Run",
        },
        Style {
          "enabled": [Function],
          "title": "Sleep",
        },
        Style {
          "enabled": [Function],
          "title": "Sport",
        },
        Style {
          "enabled": [Function],
          "title": "Walk",
        },
      ],
      "title": "themes",
    },
    Section {
      "hidden": "true",
      "styles": [
        Style {
          "enabled": [Function],
          "title": "Active",
        },
        Style {
          "enabled": [Function],
          "title": "Atmosphere",
        },
        Style {
          "enabled": [Function],
          "title": "Chant",
        },
        Style {
          "enabled": [Function],
          "title": "Children",
        },
        Style {
          "enabled": [Function],
          "title": "Coffee",
        },
        Style {
          "enabled": [Function],
          "title": "Dance",
        },
        Style {
          "enabled": [Function],
          "title": "Depress",
        },
        Style {
          "enabled": [Function],
          "title": "Deutsch",
        },
        Style {
          "enabled": [Function],
          "title": "Festival",
        },
        Style {
          "enabled": [Function],
          "title": "Happy",
        },
        Style {
          "enabled": [Function],
          "title": "History",
        },
        Style {
          "enabled": [Function],
          "title": "Holiday",
        },
        Style {
          "enabled": [Function],
          "title": "Keys",
        },
        Style {
          "enabled": [Function],
          "title": "Long",
        },
        Style {
          "enabled": [Function],
          "title": "Love",
        },
        Style {
          "enabled": [Function],
          "title": "Massage",
        },
        Style {
          "enabled": [Function],
          "title": "PlayCalimba",
        },
        Style {
          "enabled": [Function],
          "title": "PlayGuitar",
        },
        Style {
          "enabled": [Function],
          "title": "PlayPiano",
        },
        Style {
          "enabled": [Function],
          "title": "Relax",
        },
        Style {
          "enabled": [Function],
          "title": "Run",
        },
        Style {
          "enabled": [Function],
          "title": "Sleep",
        },
        Style {
          "enabled": [Function],
          "title": "Sport",
        },
        Style {
          "enabled": [Function],
          "title": "Walk",
        },
      ],
      "title": "other",
    },
  ],
  "value": [Function],
}
`;
